src/main/java/com/google/cloud/tools/app/ProcessCaller.java

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.tools.app;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.logging.Logger;
+
+/**
+ * Calls external tools like gcloud or dev_appserver. Hides system process invocation logic from
+ * {@link Action}.
+ *
+ * <p>All the logic for generating and running commands is contained in this class.
+ */
+public class ProcessCaller {
+
+  // TODO(joaomartins): Will this work in e.g., Windows?
+  private static File DEFAULT_WORKING_DIR = new File("/");
+  private static Logger LOG = Logger.getLogger(ProcessCaller.class.getName());
+  private List<String> command;
+  private File workingDirectory;
+  private boolean synchronous;
+  private String cloudSdkLocation = System.getProperty("user.home") + "/google-cloud-sdk";
+
+  private ProcessCaller(Tool tool, Collection<String> arguments) {
+    this(tool, arguments, ImmutableMap.<Option, String>of());
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags) {
+    this(tool, arguments, flags, DEFAULT_WORKING_DIR, true);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      File workingDirectory) {
+    this(tool, arguments, flags, workingDirectory, true);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      boolean synchronous) {
+    this(tool, arguments, flags, DEFAULT_WORKING_DIR, synchronous);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      File workingDirectory, boolean synchronous) {
+    this.workingDirectory = workingDirectory;
+    this.synchronous = synchronous;
+    this.command = prepareCommand(tool, arguments, flags);
+  }
+
+  public boolean call() throws GCloudExecutionException {
```

Comment:
Asynchronously, we could write logs to a file. The only thing we call asynchronously is dev_appserver, which already offers --logs_path to store request logs.

Synchronously, the library already redirects input, output and error streams to the command line.

@loosebazooka which controls are you referring to, specifically? I noticed you use a Command class and use interfaces for ProcessRunner, etc.